### PR TITLE
Decouple the video send limit and the video receive limit

### PR DESCRIPTION
**Pull request by akshay-prabhu-github**
**Issue #:** Decouple the video send and video receive limits

**Description of changes:**
Currently, the number of videos that can be sent within a meeting is coupled with the number of videos that a client can receive. This change helps us scale the video send limit independent to the video receive limit. In this change we have also added the ability for the builders to choose the remote videos that they would be interested in, so that they can page through the different remote video sources, when the number of videos being sent in the meeting are greater than the receive limit

**Testing:**
#### Unit Test

```
1929 passing (3m)

=============================== Coverage summary ===============================
Statements   : 100% ( 9775/9775 )
Branches     : 100% ( 4441/4441 )
Functions    : 100% ( 1831/1831 )
Lines        : 100% ( 9658/9658 )
================================================================================
```
#### Manual tests

Validated the following scenarios manually

1. When the number of videos being sent in the meeting are greater than the receive limit, we successfully truncate the list of videos that need to be received
2. When the client tries to receive videos beyond the video subscription limit, we encounter a signaling bad request error.

*Can these tested using a demo application? Please provide reproducible step-by-step instructions.*
Yes.

**Repro steps**
1. Request for a video send limit increase/decrease. 
2. Request for a video receive limit increase/decrease. (such that this is greater than the video receive limit)
3. Select the priority based downlink policy.
4. Check the `Paginate displayed video tiles` checkbox.
5. Join the Meeting and turn on your video.

**Expectation**
1. You should only see the number of videos that you had requested.
2. You should be able to paginate through the other set of videos.

**Checklist:**

1. Have you successfully run `npm run build:release` locally?
y

2. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
* Added an ability to choose remote video sources for `AllHighestVideoBandwidthPolicy`

3. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
Yes, we are adding a new parameter called `videoSubscriptionLimit` in our signaling protocol, that communicates to the client the maximum number of videos that a particular client can receive. This has been reviewed and approved.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/task/JoinAndReceiveIndexTask.ts [hunk 1]
```diff
@@ -71,6 +71,10 @@ export default class JoinAndReceiveIndexTask extends BaseTask {
           if (event.message.type === SdkSignalFrame.Type.JOIN_ACK) {
             // @ts-ignore: force cast to SdkJoinAckFrame
             const joinAckFrame: SdkJoinAckFrame = event.message.joinack;
+            if (joinAckFrame && joinAckFrame.videoSubscriptionLimit) {
+              videoSubscriptionLimit = joinAckFrame.videoSubscriptionLimit;
+            }
+            context.videoSubscriptionLimit = videoSubscriptionLimit;
```
Comment (ltrung) [inline src/task/JoinAndReceiveIndexTask.ts:77]: nit: We could shorten this to joinAckFrame?.videoSubscriptionLimit ? joinAckFrame.videoSubscriptionLimit : 25;

File: src/signalingclient/SignalingClientJoin.ts [hunk 2]
```diff
@@ -9,12 +9,11 @@ import ApplicationMetadata from '../applicationmetadata/ApplicationMetadata';
 export default class SignalingClientJoin {
   /** Initializes a SignalingClientJoin with the given properties.
    *
-   * @param maxVideos The maximum number of video tiles to send.
    * @param sendBitrates Whether the server should send Bitrates messages.
    * @param applicationMetadata [[ApplicationMetadata]].
    */
+
```
Comment (ltrung) [inline src/signalingclient/SignalingClientJoin.ts:15]: a bit of a stretch but since this class is public, we cannot really delete this param.

File: src/videodownlinkbandwidthpolicy/VideoAdaptiveProbePolicy.ts [hunk 3]
```diff
@@ -49,14 +49,4 @@ export default class VideoAdaptiveProbePolicy extends VideoPriorityBasedPolicy {
       this.videoPreferences = undefined;
     }
   }
-
-  /**
-   * [[VideoAdaptiveProbePolicy]] does not allow setting video preferences and this function
-   * will be a no-op.  Please use [[VideoPriorityBasedPolicy]] directly if you would like to set
-   * preferences.
-   */
-  chooseRemoteVideoSources(_preferences: VideoPreferences): void {
```
Comment (ltrung) [inline src/videodownlinkbandwidthpolicy/VideoAdaptiveProbePolicy.ts:58]: Adaptive auto manage the preference so this is a change in behavior. Should we just truncate here instead?

File: src/videostreamidset/DefaultVideoStreamIdSet.ts [hunk 4]
```diff
@@ -66,4 +66,9 @@ export default class DefaultVideoStreamIdSet implements VideoStreamIdSet {
   toJSON(): number[] {
     return this.array();
   }
+
+  truncate(length: number): DefaultVideoStreamIdSet {
```
Comment (ltrung) [inline src/videostreamidset/DefaultVideoStreamIdSet.ts:70]: Should the return type be VideoStreamIdSet?

File: src/videostreamidset/DefaultVideoStreamIdSet.ts [hunk 5]
```diff
@@ -66,4 +66,9 @@ export default class DefaultVideoStreamIdSet implements VideoStreamIdSet {
...
+    const x: number[] = this.array();
+    return new DefaultVideoStreamIdSet(x.splice(0, length));
```
Comment (ltrung) [inline src/videostreamidset/DefaultVideoStreamIdSet.ts:72]: Should we log a warning that we will truncate since there are more video to subscribe than the limit number?

File: src/videodownlinkbandwidthpolicy/AllHighestVideoBandwidthPolicy.ts [hunk 6]
```diff
@@ -15,6 +16,9 @@ import VideoDownlinkBandwidthPolicy from './VideoDownlinkBandwidthPolicy';
 export default class AllHighestVideoBandwidthPolicy implements VideoDownlinkBandwidthPolicy {
   private optimalReceiveSet: VideoStreamIdSet;
   private subscribedReceiveSet: VideoStreamIdSet;
+  private areVideoSourcesChosen: boolean;
+  private videoSources: VideoSource[];
```
Comment (ltrung) [inline src/videodownlinkbandwidthpolicy/AllHighestVideoBandwidthPolicy.ts:20]: Can we just do videoSources to be VideoSources[]| undefined? Below we can just check whether videoSources !== undefined and not need `areVideoSourcesChosen` variable.

File: src/videodownlinkbandwidthpolicy/VideoAdaptiveProbePolicy.ts (hunk 3 above)
Comment (akshay-prabhu-github) [reply src/videodownlinkbandwidthpolicy/VideoAdaptiveProbePolicy.ts:58]: I accidentally deleted that function in VideoAdaptivePolicy. This class needs to stay, else it will be similar to PriorityBasedDL.

And yes, we should just truncate here.

File: src/videodownlinkbandwidthpolicy/AllHighestVideoBandwidthPolicy.ts [hunk 7]
```diff
@@ -23,10 +27,17 @@ export default class AllHighestVideoBandwidthPolicy implements VideoDownlinkBand
...
+    this.areVideoSourcesChosen = false;
   }
 
   updateIndex(videoIndex: VideoStreamIndex): void {
-    this.optimalReceiveSet = this.calculateOptimalReceiveSet(videoIndex);
+    this.videoIndex = videoIndex;
+    if (!this.areVideoSourcesChosen) {
+      this.optimalReceiveSet = this.calculateOptimalReceiveSet(videoIndex);
+    } else {
+      this.chooseRemoteVideoSources(this.videoSources);
```
Comment (ltrung) [inline src/videodownlinkbandwidthpolicy/AllHighestVideoBandwidthPolicy.ts:39]: Instead of this, can we:
- only set videoSources in `chooseRemoteVideoSources`
- in `calculateOptimalreceiveSet`, calculate the set and factor in videoSources.

File: src/videodownlinkbandwidthpolicy/VideoAdaptiveProbePolicy.ts (hunk 3 above)
Comment (ltrung) [reply src/videodownlinkbandwidthpolicy/VideoAdaptiveProbePolicy.ts:58]: Also updateIndex already update the preference (which one happens first? updateIndex or remoteVideoSourcesDidChange do you know ?) so there is a conflict here.

**Review by ltrung — COMMENTED** [review]
We need to update our guides about these changes I assume?

**Review by ltrung — COMMENTED** [review]
Should we throw error or log error in priority based policy and other if they try to set more video sources than the limit?

File: src/videostreamidset/DefaultVideoStreamIdSet.ts (hunk 5 above)
Comment (akshay-prabhu-github) [reply src/videostreamidset/DefaultVideoStreamIdSet.ts:72]: good point, given this is a library function, we need not log it here. 

But I will log it in ReceiveVideoStreamIndexTask.ts before we actually truncate it.

File: src/videostreamidset/DefaultVideoStreamIdSet.ts (hunk 4 above)
Comment (akshay-prabhu-github) [reply src/videostreamidset/DefaultVideoStreamIdSet.ts:70]: We are returning the `DefaultVideoStreamIdSet` in the clone function. so to maintain conformity, I am using the same signature here.

File: src/videodownlinkbandwidthpolicy/AllHighestVideoBandwidthPolicy.ts (hunk 6 above)
Comment (akshay-prabhu-github) [reply src/videodownlinkbandwidthpolicy/AllHighestVideoBandwidthPolicy.ts:20]: will do

File: src/signalingclient/SignalingClientJoin.ts (hunk 2 above)
Comment (akshay-prabhu-github) [reply src/signalingclient/SignalingClientJoin.ts:15]: ohh interesting. I can add a `@deprecated` parameter to it and log an error in-case anyone tries to use that parameter.

File: src/videodownlinkbandwidthpolicy/AllHighestVideoBandwidthPolicy.ts (hunk 7 above)
Comment (akshay-prabhu-github) [reply src/videodownlinkbandwidthpolicy/AllHighestVideoBandwidthPolicy.ts:39]: Good point. Let me give it a try. I feel that approach should also work and might be cleaner even.

File: src/task/JoinAndReceiveIndexTask.ts (hunk 1 above)
Comment (akshay-prabhu-github) [reply src/task/JoinAndReceiveIndexTask.ts:77]: I tried doing this and the single line went too long. so would like to continue to keep it as is.

File: src/videodownlinkbandwidthpolicy/VideoAdaptiveProbePolicy.ts (hunk 3 above)
Comment (akshay-prabhu-github) [reply src/videodownlinkbandwidthpolicy/VideoAdaptiveProbePolicy.ts:58]: 1. The UpdateIndex happens first and it provides all the new index values, which will be used by the `chooseSubscription` function.
2. The `chooseSubscription` function is called in the `ReceiveVideoStreamIndexTask` in the `resubscribe` function
3. In the `resubscribe` function we truncate the list to have only 25 videos in the `videoToReceive` list
4. Then finally the remoteVideoSourcesDidChange function is called incase, anything has changed.

File: CHANGELOG.md [hunk 8]
```diff
@@ -5,6 +5,18 @@ All notable changes to this project will be documented in this file.
...
+### Added
+- Ability to choose remote video sources in `AllHighestVideoBandwidthPolicy`
+
+### Removed
+
+### Fixed
+
+### Changed
+- Enforced a video receive limit incase the number of videos shared in the meeting
+are greater than the limit.
```
Comment (hensmi-amazon) [inline CHANGELOG.md:18]: Nit: I think we've settled on just using word wrap here :) .

Comment (hensmi-amazon) [reply CHANGELOG.md:18]: Also can we mention that this is currently 25, but may change in the future?

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,18 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+### Added
+- Ability to choose remote video sources in `AllHighestVideoBandwidthPolicy`
```
Comment (hensmi-amazon) [inline CHANGELOG.md:10]: I'd link to the guide here.

File: src/audiovideocontroller/AudioVideoControllerState.ts [hunk 10]
```diff
@@ -106,6 +106,9 @@ export default class AudioVideoControllerState {
   // itself correspond to transceivers; 0 in this list corresponds to an inactive tranceiver.
   videoSubscriptions: number[] | null = null;
 
+  // Total number of videos streams that can be subscribed to.
+  videoSubscriptionLimit: number = 0;
```
Comment (hensmi-amazon) [inline src/audiovideocontroller/AudioVideoControllerState.ts:110]: Nit: Comment where this number comes from (because it will always and only come from the backend), that it may change, and where it is used (particularly since `AudioVideoControllerState` is already sort of a global store, and easily abusible; if possible, we at least want to warn future abusers of this value about how it is set).

File: src/signalingclient/DefaultSignalingClient.ts [hunk 11]
```diff
@@ -83,7 +83,6 @@ export default class DefaultSignalingClient implements SignalingClient {
     this.logger.info('sending join');
     const joinFrame = SdkJoinFrame.create();
     joinFrame.protocolVersion = 2;
-    joinFrame.maxNumOfVideos = settings.maxVideos;
```
Comment (hensmi-amazon) [inline src/signalingclient/DefaultSignalingClient.ts:86]: End of a (magic number) era.

File: src/signalingclient/SignalingClientJoin.ts [hunk 12]
```diff
@@ -9,10 +9,12 @@ import ApplicationMetadata from '../applicationmetadata/ApplicationMetadata';
 export default class SignalingClientJoin {
   /** Initializes a SignalingClientJoin with the given properties.
    *
-   * @param maxVideos The maximum number of video tiles to send.
+   * 
+   * @param maxVideos The maximum number of video tiles to send. // deprecated
```
Comment (hensmi-amazon) [inline src/signalingclient/SignalingClientJoin.ts:13]: Can we say why its deprecated. I guess not a big deal since this is more internal but should be be no-op-ing it and resetting to default?  Some builders probably still have this hard coded in their fork.

File: src/task/JoinAndReceiveIndexTask.ts [hunk 13]
```diff
@@ -37,6 +37,7 @@ export default class JoinAndReceiveIndexTask extends BaseTask {
   async run(): Promise<void> {
     const indexFrame = await new Promise<SdkIndexFrame>((resolve, reject) => {
       const context = this.context;
+      let defaultVideoSubscriptionLimit: number = 25;
```
Comment (hensmi-amazon) [inline src/task/JoinAndReceiveIndexTask.ts:40]: Feel like this should just be set as default in `AudioVideoControllerState`? the possibility of it being 0 makes me a bit uncomfortable.

File: src/task/ReceiveVideoStreamIndexTask.ts [hunk 14]
```diff
@@ -164,7 +164,15 @@ export default class ReceiveVideoStreamIndexTask
...
     }
 
-    this.context.videosToReceive = videoDownlinkBandwidthPolicy.chooseSubscriptions();
+    const videoSubscriptionIdSet = videoDownlinkBandwidthPolicy.chooseSubscriptions();
+    this.context.videosToReceive = videoSubscriptionIdSet.truncate(
+      this.context.videoSubscriptionLimit
+    );
+
+    if (videoSubscriptionIdSet.size() > this.context.videosToReceive.size()) {
+      this.logger.warn(`video receive limit exceeded. Limiting the videos to ${this.context.videosToReceive.size()}`);
```
Comment (hensmi-amazon) [inline src/task/ReceiveVideoStreamIndexTask.ts:173]: Can we make this log a bit more verbose/helpful and make it clear that builders should change their application code. Anecdotally, some times builders see a log like this and think its a bug on our end.

Comment (hensmi-amazon) [reply src/task/ReceiveVideoStreamIndexTask.ts:173]: Seems like we should recommend against using AdaptiveProbe? since that will run into this log quite often?

**Review by hensmi-amazon — COMMENTED** [review]
looking good!

File: src/signalingclient/SignalingClientJoin.ts (hunk 2 above)
Comment (ltrung) [reply src/signalingclient/SignalingClientJoin.ts:17]: Could we document how it works before and now what would be and why we deprecate it?

File: src/task/JoinAndReceiveIndexTask.ts (hunk 13 above)
Comment (ltrung) [reply src/task/JoinAndReceiveIndexTask.ts:40]: Agree

File: src/task/ReceiveVideoStreamIndexTask.ts (hunk 14 above)
Comment (ltrung) [reply src/task/ReceiveVideoStreamIndexTask.ts:173]: imo we should just support in priority-based downlink policy but i guess we can't as priority-based does not work in FF :(

File: CHANGELOG.md (hunk 9 above)
Comment (akshay-prabhu-github) [reply CHANGELOG.md:10]: I have currently linked to https://aws.github.io/amazon-chime-sdk-js/modules/videolayout.html#downlink-policy. Will work with you to update that section of the document.

File: src/signalingclient/SignalingClientJoin.ts (hunk 12 above)
Comment (akshay-prabhu-github) [reply src/signalingclient/SignalingClientJoin.ts:13]: > Can we say why its deprecated

Yes, I can add a reason for the deprecation.

> I guess not a big deal since this is more internal but should be be no-op-ing it and resetting to default?
Yes, it will be resetting to default.

File: src/task/JoinAndReceiveIndexTask.ts (hunk 13 above)
Comment (akshay-prabhu-github) [reply src/task/JoinAndReceiveIndexTask.ts:40]: Yeah, I can have it set to default as 25 in the proto and the AudioVideoControllerState.

File: src/task/ReceiveVideoStreamIndexTask.ts (hunk 14 above)
Comment (akshay-prabhu-github) [reply src/task/ReceiveVideoStreamIndexTask.ts:173]: Yeah, I can have it set to default as 25 in the protobuf and the AudioVideoControllerState

Comment (akshay-prabhu-github) [reply src/task/ReceiveVideoStreamIndexTask.ts:173]: Updated the comment with additional message.

```
video receive limit exceeded. Limiting the videos to 25.
Please consider using AllHighestVideoBandwidthPolicy or VideoPriorityBasedPolicy along 
with choosRemoteVideoSources api to select the video sources to be displayed.
```

File: src/signalingclient/SignalingClientJoin.ts (hunk 2 above)
Comment (akshay-prabhu-github) [reply src/signalingclient/SignalingClientJoin.ts:18]: Updated the comments

**Review by hensmi-amazon — APPROVED** [review]
🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

File: guides/16_Video_Layout.md [hunk 15]
```diff
@@ -48,7 +48,7 @@ The downlink policy controls which remote video sources are received downstream
 
 ##### AllHighestVideoBandwidthPolicy (Default)
 
-This policy subscribes to the highest quality video layer of all publishers. Under constrained network, some or all videos may freeze as packets are lost on the network, and they will recover automatically once network congestion is reduced.
+This policy subscribes to the highest quality video layer of all publishers. This policy allows builders to choose which remote video sources to subscribe to through the use of `chooseRemoteVideoSources` api.  Under constrained network, some or all videos may freeze as packets are lost on the network, and they will recover automatically once network congestion is reduced.
```
Comment (ltrung) [inline guides/16_Video_Layout.md:51]: nit: is it API or api? I think we use API in our documentation.

Comment (akshay-prabhu-github) [reply guides/16_Video_Layout.md:51]: ohh ok. Will update it to conform to the standard.

**Review by hensmi-amazon — APPROVED** [review]

**Comment by hensmi-amazon** [comment]
Funny how after dismissing stale reviews it just looks like I'm commenting a bunch of rocket emojis for no reason ;)
